src/main/relation/join_relation.cpp

```diff
@@ -34,6 +37,104 @@ unique_ptr<QueryNode> JoinRelation::GetQueryNode() {
 	return std::move(result);
 }
 
+static string GetDeduplicatedName(case_insensitive_map_t<string> &bindings, const string &name) {
+	auto it = bindings.find(name);
+	if (it == bindings.end()) {
+		// Binding not seen, safe to assume it's unique (I assume this errors later anyways?)
+		return name;
+	}
+
+	// If 'second' is not empty, this is the start of a deduplication chain that we need to follow
+	while (!it->second.empty()) {
+		it = bindings.find(it->second);
+	}
+	return it->first;
+}
+
+static void ReplaceQualificationRecursive(unique_ptr<ParsedExpression> &expr,
+                                          case_insensitive_map_t<string> &changed_bindings) {
+	if (expr->GetExpressionType() == ExpressionType::COLUMN_REF) {
+		auto &col_ref = expr->Cast<ColumnRefExpression>();
+		if (col_ref.IsQualified()) {
+			auto deduplicated_name = GetDeduplicatedName(changed_bindings, col_ref.GetTableName());
+			col_ref.ReplaceOrRemoveTableName(deduplicated_name);
+		}
+	} else {
+		ParsedExpressionIterator::EnumerateChildren(*expr, [&changed_bindings](unique_ptr<ParsedExpression> &child) {
+			ReplaceQualificationRecursive(child, changed_bindings);
+		});
+	}
+}
+
+static void AddBinding(case_insensitive_map_t<string> &bindings, string &name) {
+	auto it = bindings.find(name);
+	if (it == bindings.end()) {
+		bindings[name] = "";
+		return;
+	}
+	// If the value is not an empty string, the binding name already has a duplicate
+	// this creates a chain of name -> name_1 -> name_1_1 -> ...
+	while (!it->second.empty()) {
```

Comment:
The reason this is more difficult than `DeduplicateColumns` is because in `GetDeduplicatedName` we need to look up the `old -> new` transformation.

If we for example have `[old, old_1, old]` this will be deduplicated to (replacing old->new for illustration purposes) `[new, new_1, new_2]`
If we look up `old_1` we won't know if this has been mapped to `new_1` or `new_2`

With the current solution we'll know because `old_1` will have its own entry ....

----------------------------

As I am typing this I realize that the current solution suffers from the same problem:
```c++
	string GetDeduplicatedName(const string &name) {
		auto it = bindings.find(name);
		if (it == bindings.end()) {
			// Binding not seen, safe to assume it's unique (I assume this errors later anyways?)
			return name;
		}

		// If 'second' is not empty, this is the start of a deduplication chain that we need to follow
		while (!it->second.empty()) {
			it = bindings.find(it->second);
		}
		return it->first;
	}
```

This will always return the last link of the chain, returning `new_2` for both the second `old` and `old_1`.
Though perhaps this works out fine because of the order in which the method is called?

Reply:
I'll write a test for this